sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,461 @@
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_GRAPH` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== SYCL Graph Terminology
+
+Table 2. Terminology.
+[%header,cols="1,3"]
+|===
+|Concept|Description
+|graph| Class that stores structured work units and their dependencies
+|node| The unit of work. Can have different attributes.
+|edge| Dependency between work units. Happens-before relation.
+|===
+
+== Node
+
+Node is a class that encapsulates tasks like SYCL kernel functions or host tasks for deferred execution.
+A graph has to be created first, the structure of a graph is defined second by adding nodes and edges.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  class node{
+  };
+}
+----
+
+== Edge
+
+A dependency between two nodes representing a happens-before relationship. `sender` and `receiver` may be associated to different graphs.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+	// Adding dependency between two nodes.
+    	void make_edge(node sender, node receiver);
+}
+----
+
+== Graph
+
+Graph is a class that represents a directed acyclic graph of nodes. 
+A graph can have different states, can be nested, can have multiple root nodes that are scheduled for execution first and multiple leaf nodes that are scheduled for execution last. The execution of a graph has been completed when all leaf node tasks have been completed.
+Member functions as listed in Table 3 to 6 can be used to add nodes to a graph.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  enum class graph_state{
+    modifiable,
+    executable
+  };
+
+  template<graph_state State>
+  class graph {
+  public:
+    operator graph<graph_state::executable>();
+  };
+  
+  graph<graph_state::modifiable> make_graph();
+  
+  graph<graph_state::executable> compile(const graph<graph_state::modifiable> Graph);
+  
+}
+
+----
+
+The following member functions are added to the queue class.
+
+[source,c++]
+----
+
+namespace sycl {
+
+event submit(const ext::oneapi::experimental::graph<ext::oneapi::experimental::graph_state::executable>& my_graph);
+
+} // namespace sycl
+
+----
+
+=== Executable Graph
+
+A `graph` object in `graph_state::executable` represents a user generated device and context specific execution object that is submitted to a queue for execution.
+The structure of such a `graph` object in this state is immutable and cannot be changed, so are the tasks assigned with each node.
```

Comment:
I like the idea of encoding the immutability in the type system.
But at the same time it means 2 different kinds of graph. Of course the conversion operator you propose could just share the underlying graph structure underneath to reduce the cost of the copy.
But what happens if the first modifiable graph is later updated? You do not want to change the executable graph.
Do we want a copy-on-write? Or just a move semantics into the executable graph and resetting the modifiable graph?
Otherwise, just 1 graph type with a runtime state representing executable/modifiable and having the member functions guarded by this state so they can throw if something is forbidden?